### PR TITLE
Allow for larger iCal exports

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: https://central.tri.be/issues/65037

This PR:

* refactors the `iCal` class to use the container and be instance based and not static based
* implements a way for developers to control the feed export size independently of the `posts_per_page` option and the current view option
* tests!

### CODE REVIEW

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $template is never used | [iCal.php:124](https://github.com/moderntribe/the-events-calendar/blob/715433248c3eab6e58e59dd949775e8a9db02ef7/src/Tribe/iCal.php#L124)
:warning: | Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$content' | [iCal.php:373](https://github.com/moderntribe/the-events-calendar/blob/715433248c3eab6e58e59dd949775e8a9db02ef7/src/Tribe/iCal.php#L373)
:x: | There must be a single space before the last parenthesis | [ical.php:13](https://github.com/moderntribe/the-events-calendar/blob/715433248c3eab6e58e59dd949775e8a9db02ef7/src/functions/template-tags/ical.php#L13)
:x: | There must be a single space before the last parenthesis | [ical.php:28](https://github.com/moderntribe/the-events-calendar/blob/715433248c3eab6e58e59dd949775e8a9db02ef7/src/functions/template-tags/ical.php#L28)

> via jenkins/codesniffer-phpcs

File: src/Tribe/iCal.php [hunk 1]
```diff
@@ -218,11 +201,29 @@ public static function generate_ical_feed( $post = null ) {
...
+			$count = apply_filters( 'tribe_ical_feed_posts_per_page', $this->feed_default_export_count );
+
+			$count = is_numeric( $count ) && is_int( $count ) && $count > 0 ? $count : $this->feed_default_export_count;
+
+			/** @var WP_Query $wp_query */
+			global $wp_query;
+
+			$query_posts_per_page = $wp_query->get( 'posts_per_page' );
+			if ( $count > $query_posts_per_page ) {
+				$query       = new WP_Query( array_merge( $wp_query->query, array( 'posts_per_page' => $count ) ) );
```
Comment (bordoni) [inline src/Tribe/iCal.php]: Weird spacing.

File: src/Tribe/iCal.php [hunk 2]
```diff
@@ -359,9 +360,31 @@ public static function generate_ical_feed( $post = null ) {
 		$content = apply_filters( 'tribe_ical_properties', $content );
 		$content .= $events;
 		$content .= 'END:VCALENDAR';
-		echo $content;
 
-		exit;
+		if ( $echo ) {
+			echo $content;
```
Comment (bordoni) [inline src/Tribe/iCal.php]: Why not `exit( $content );`

File: src/Tribe/iCal.php [hunk 3]
```diff
@@ -218,11 +201,29 @@ public static function generate_ical_feed( $post = null ) {
...
+			 * Filters the number of upcoming events the iCal feed should export.
+			 *
+			 * This filter allows developer to override the pagination setting and the default value
+			 * to export a number of events that's inferior or superior to the one shown on the page.
+			 * The minimum value is 1.
+			 *
+			 * @param int $count The number of upcoming events that should be exported in the
+			 *                   feed, defaults to 30.
+			 */
+			$count = apply_filters( 'tribe_ical_feed_posts_per_page', $this->feed_default_export_count );
```
Comment (bordoni) [inline src/Tribe/iCal.php]: Can you move this filter towards the method? `get_feed_default_export_count`, to ensure that every single time someone fetches that value we have it be filtered.

File: src/Tribe/iCal.php [hunk 4]
```diff
@@ -218,11 +201,29 @@ public static function generate_ical_feed( $post = null ) {
...
+
+			$count = is_numeric( $count ) && is_int( $count ) && $count > 0 ? $count : $this->feed_default_export_count;
```
Comment (bordoni) [inline src/Tribe/iCal.php]: Lets avoid using the prop `$this-> feed_default_export_count`, and use the method instead?

File: src/Tribe/iCal.php [hunk 5]
```diff
@@ -359,9 +360,31 @@ public static function generate_ical_feed( $post = null ) {
 		$content = apply_filters( 'tribe_ical_properties', $content );
 		$content .= $events;
 		$content .= 'END:VCALENDAR';
-		echo $content;
 
-		exit;
+		if ( $echo ) {
+			echo $content;
+			exit;
```
Comment (barryhughes) [inline src/Tribe/iCal.php]: Should we be using `tribe_exit()` as a general rule now?

File: tests/integration/Tribe/Events/iCalTest.php [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+		/** @var \WP_Query $wp_query */
+		global $wp_query;
+
+		// run the query, simulates what would happen on the page
+		$wp_query = new \WP_Query( [ 'post_type' => $post_type, 'posts_per_page' => $posts_per_page ] );
+
+		add_filter(
+			'tribe_ical_feed_posts_per_page', function () use ( $filter_count ) {
+			return $filter_count;
+		}
```
Comment (barryhughes) [inline tests/integration/Tribe/Events/iCalTest.php]: Spacing/indentation is weird here. How about:

```
add_filter(
	'tribe_ical_feed_posts_per_page', function () use ( $filter_count ) {
		return $filter_count;
	}
);
```

(Or even have the closing function brace and parethensis on the same line, like `} );`?

File: src/Tribe/iCal.php (hunk 3 above)
Comment (lucatume) [inline src/Tribe/iCal.php]: The idea is to use the filter where people might be looking for it. The method `get_feed_default_export_count` is an accessor of the class property, not of the value that will be used in the end. That's used in tests to distinguish between the two (filtered and non-filtered value). Why could we want to have and keep track of filtered and unfiltered values? If we ever make an option of this in Event Settings that unfiltered value is our "Reset to defaults" information written in code; plus we know we have a default we can trust.

File: src/Tribe/iCal.php (hunk 4 above)
Comment (lucatume) [inline src/Tribe/iCal.php]: See above. We are falling back to a default value we control and trust without diminishing its filterability.

File: src/Tribe/iCal.php (hunk 2 above)
Comment (lucatume) [inline src/Tribe/iCal.php]: copied and pasted really, joining this suggestion and Barry's one about the use of `tribe_exit`

File: src/Tribe/iCal.php (hunk 1 above)
Comment (lucatume) [inline src/Tribe/iCal.php]: yep, corrected

**Comment by barryhughes** [comment]
:cow:

**Pull request merged**